### PR TITLE
feat: add additional manifest validations

### PR DESCRIPTION
**Pull request by peusebiu**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by rchincha** [comment]
2023-07-11T16:32:58.4637139Z Will run [1m79[0m of [1m79[0m specs
2023-07-11T16:32:58.5333546Z 
2023-07-11T16:32:58.5334539Z     [38;5;13mWARNING: image manifest with no layers is not supported
2023-07-11T16:32:58.5334962Z     [0m
2023-07-11T16:32:58.5336781Z     /home/runner/work/zot/zot/distribution-spec/conformance/02_push_test.go:366

**Comment by rchincha** [comment]
image-spec is completely independent and the upstream schema validation requires a minimum of one layer

We happen to use image-spec for our layout but dist-spec allows zero layers, which we should also allow.

This is the conflicting requirement.

File: pkg/storage/common/common.go [hunk 1]
```diff
@@ -71,27 +73,36 @@ func ValidateManifest(imgStore storageTypes.ImageStore, repo, reference, mediaTy
 	switch mediaType {
 	case ispec.MediaTypeImageManifest:
 		var manifest ispec.Manifest
+
+		// validate manifest
+		if err := schema.ValidatorMediaTypeManifest.Validate(bytes.NewBuffer(body)); err != nil {
```
Comment (rchincha) [inline pkg/storage/common/common.go:78]: If the "err" says that there is only ONE error and that is missing layers, then we should still pass.

Comment (rchincha) [reply pkg/storage/common/common.go:78]: See the "warning" in the comments. We want to support that case in `zot`

**Comment by rchincha** [comment]
FYI, https://github.com/opencontainers/distribution-spec/pull/447

File: pkg/storage/common/common.go [hunk 2]
```diff
@@ -777,6 +771,19 @@ func IsNonDistributable(mediaType string) bool {
 		mediaType == ispec.MediaTypeImageLayerNonDistributableZstd //nolint:staticcheck
 }
 
+func IsEmptyLayersError(err error) bool {
+	var validationErr schema.ValidationError
+	if errors.As(err, &validationErr) {
+		if len(validationErr.Errs) == 1 && strings.Contains(err.Error(), "layers: Array must have at least 1 items") {
```
Comment (andaaron) [inline pkg/storage/common/common.go:777]: Define this error separately. Also can we add a test (if we don't have one already) where there are multiple validation errors, and this is just 1 of them?

Comment (peusebiu) [reply pkg/storage/common/common.go:777]: added in TestArtifact in controller_test.go

what do you mean to define it separately? it's defined in schema package

Comment (andaaron) [reply pkg/storage/common/common.go:777]: I see a hardcoded string `"layers: Array must have at least 1 items"`.
That's what I meant.

File: pkg/api/controller_test.go [hunk 3]
```diff
@@ -5557,7 +5557,86 @@ func TestArtifactReferences(t *testing.T) {
...
+					Digest:    digest,
+					Size:      int64(len(mcontent)),
+				})
+
+				mcontent, err = json.Marshal(manifest)
+				So(err, ShouldBeNil)
+				resp, err = resty.R().SetHeader("Content-Type", ispec.MediaTypeImageManifest).
+					SetBody(mcontent).Put(baseURL + fmt.Sprintf("/v2/%s/manifests/1.0", repoName))
+				So(err, ShouldBeNil)
+				// modified from http.StatusBadRequest tot http.StatusCreated
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/api/controller_test.go:5817]: ok

**Review by andaaron — APPROVED** [review]

**Pull request merged**